Implement slice input fusion.

trentlo opened the pull request:
Implement code generation for input-fusible slices. It is useful to support split-like semantic. Before this change, slices can only be fused into their successors. This change enables the slices to be fused with their predecessors.

trentlo committed:
Implement slice input fusion.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc:
```diff
@@ -234,6 +234,37 @@ bool IsReductionFromOrToContiguousDimensions(const HloInstruction& reduce) {
+    return absl::c_all_of(strides, [](int stride) { return stride == 1; });
+  };
+
+  const auto root = unnested_hlo.fused_expression_root();
```
Could we spell out the type explicitly here?

trentlo replied:
Done.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emission_utils.h:
```diff
@@ -157,6 +157,11 @@ bool ImplementedAsLibraryCall(const HloInstruction& hlo);
 
+// Whether it is an input fusion whose root is either a non-strided slice
+// or a tuple of non-strided slices.
+bool IsInputFusibleSlices(const HloInstruction& unnested_hlo,
```
Could you document the semantics of `verify_no_strides`?

trentlo replied:
No problem.

cheshire replied:
"verify" is not clear, how about something like

`Returns whether input is either an input fusion [...]. If verify_no_strides is true, returns false unless ...`

trentlo replied:
Good suggestion. Taken.

cheshire replied:
Sorry for nitpicking, now the first two sentences almost duplicate each other. Can we write "Returns ..." with all the relevant information available after?

trentlo replied:
My oversight. Will do.

trentlo replied:
Done.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc:
```diff
@@ -234,6 +234,37 @@ bool IsReductionFromOrToContiguousDimensions(const HloInstruction& reduce) {
+    return false;
+  }
+
+  bool all_non_strided_slices =
```
Can we simply return the result of `c_all_of`?

trentlo replied:
Done.

cheshire replied:
I've meant `return absl::c_all_of(...);`?

trentlo replied:
Done. I know what you meant. I was just thinking that the return name also gives some semantics.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -301,6 +301,41 @@ llvm::Type* GetIndexTypeForKernel(const HloInstruction* hlo, int64 launch_size,
+        "Only non-strided slices are supported.");
+  }
+
+  auto& root = *fusion.fused_expression_root();
```
Explicit type.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -301,6 +301,41 @@ llvm::Type* GetIndexTypeForKernel(const HloInstruction* hlo, int64 launch_size,
   return b->getInt32Ty();
 }
 
+// Check whether all the shapes of the slice input operands are the same.
```
Could you clarify the contract here? It's not just "checking", since we are also returning the shape.

Something like "Returns FailedPrecondition if input shapes differ, otherwise, returns the shape of ..."

trentlo replied:
Revised the name and added some more comments.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -301,6 +301,41 @@ llvm::Type* GetIndexTypeForKernel(const HloInstruction* hlo, int64 launch_size,
+    return root.operands()[0]->shape();
+  }
+
+  DCHECK(root.opcode() == HloOpcode::kTuple);
```
I would prefer just to `CHECK`.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+  VLOG(10) << "Emitting slice input fusion for " << unnested_hlo->ToString();
+
+  HloInstruction* slice_or_tuple = unnested_hlo->fused_expression_root();
+  auto slice_instructions = ([&]() -> absl::Span<HloInstruction* const> {
```
Two pairs of parens are not required: `auto x = [&] { ... }()` also works.

trentlo replied:
Good suggestion.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+    if (slice_or_tuple->opcode() == HloOpcode::kSlice) {
+      return absl::Span<HloInstruction* const>(&slice_or_tuple, 1);
+    }
+    DCHECK(slice_or_tuple->opcode() == HloOpcode::kTuple);
```
CHECK

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+  FusedIrEmitter fused_emitter(GetGeneratorForOperandIrArrays(unnested_hlo),
+                               &elem_emitter);
+  TF_CHECK_OK(unnested_hlo->fused_expression_root()->Accept(&fused_emitter));
+  for (size_t i = 0; i < slice_instructions.size(); ++i) {
```
For-each also works?

trentlo replied:
Good suggestion.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+
+  // Emit for slice_instructions begins here.
+  llvm::BasicBlock* init_block = b_.GetInsertBlock();
+  llvm::BasicBlock* exit_block;
```
Initialize on the same line?

trentlo replied:
Good suggestion.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+  // First, generate blocks and guard_blocks for each slice. Will populate
+  // them below.
+  std::vector<llvm::BasicBlock*> emitted_guard_blocks;
+  std::vector<llvm::BasicBlock*> emitted_slice_blocks;
```
Optional: it's customary to do `.reserve(slice_instructions.size())` in such cases.

trentlo replied:
Added.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+    // guarding_cond := index >= start && index < limit.
+    llvm::Value* guarding_cond = nullptr;
+    for (size_t dim = 0; dim < slice->slice_starts().size(); ++dim) {
+      DCHECK(slice->slice_strides(dim) == 1);
```
CHECK

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+  // Jump from init_block to the first guard_block.
+  init_block->getTerminator()->eraseFromParent();
+  b_.SetInsertPoint(init_block);
+  Br(emitted_guard_blocks[0]);
```
High-level comment: no other methods in this class use such low-level primitives.

I don't completely understand the control flow here, but is it possible to write it using higher-level constructs such as ksl.If ? And if not, could we add a comment why not?

trentlo replied:
A similar style can be found in the codegen of Concat here:
https://github.com/tensorflow/tensorflow/blob/483c16701b0b98e4de58f3041c20cc1886069b3a/tensorflow/compiler/xla/service/elemental_ir_emitter.cc#L1742

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+    for (size_t dim = 0; dim < slice->slice_starts().size(); ++dim) {
+      DCHECK(slice->slice_strides(dim) == 1);
+      auto larger_or_equal_than_start =
+          ICmpSGE(index.multidim()[dim],
```
Nitpick: could we use full names without shortcuts, e.g. `b_.CreateICmpSGE`? Full names work much better with tooling.

trentlo replied:
Done.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+      auto smaller_than_limit =
+          ICmpSLT(index.multidim()[dim],
+                  index.GetConstantWithIndexType(slice->slice_limits(dim)));
+      auto within_range = And(larger_or_equal_than_start, smaller_than_limit);
```
Full type in both cases.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+                  index.GetConstantWithIndexType(slice->slice_limits(dim)));
+      auto within_range = And(larger_or_equal_than_start, smaller_than_limit);
+      guarding_cond = (guarding_cond == nullptr)
+                          ? within_range
```
It seems cleaner to collect all guards in a vector first, and then use `b_.CreateAnd(conds)`

trentlo replied:
Good suggestion.

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc:
```diff
@@ -0,0 +1,159 @@
+namespace gpu {
+namespace {
+
+class GpuSliceInputFusionTest : public GpuCodegenTest {
```
These are execution and HLO tests.
Given that this change affects codegen function, could we also have a FileCheck-like test for the produced IR?

cheshire commented:
Great, thanks!

For the commit description it would be awesome to also have one micro-benchmark with `--xla_hlo_profile`before and after.

cheshire commented on tensorflow/compiler/xla/service/gpu/tests/gpu_input_fusible_slice_test.cc:
```diff
@@ -0,0 +1,159 @@
+          .ValueOrDie();
+  CompileAndVerifyIr(std::move(hlo_module),
+                     R"(
+; CHECK-LABEL: define void @fusion
```
Could we have a slightly more involved codegen test? This commit adds quite a lot of codegen logic.

trentlo replied:
Do you mean to add a more involved test pattern or you feel the IR CHECK is not enough?

cheshire replied:
Yes, a more involved pattern would be nice.

trentlo replied:
I gave some more contexts in the main discussion thread. I will wait for your response there.

trentlo committed:
Addressed review comments.

trentlo reviewed (commented):
Addressed all the comments except the one asking for more tests.

trentlo commented:
> Great, thanks!
> 
> For the commit description it would be awesome to also have one micro-benchmark with `--xla_hlo_profile`before and after.

Thanks for the review.

No problem to add a micro-benchmark. Could you give me some pointers about how to go that? Any example?

trentlo committed:
Typo fix and minor twist.

cheshire commented:
> No problem to add a micro-benchmark. Could you give me some pointers about how to do that? Any example?

Thanks, e.g.
https://github.com/tensorflow/tensorflow/commit/c6ccf6316430193e18dde016698d2d8c9989b30a

You can get the numbers by running with `XLA_FLAGS=--xla_hlo_profile`

Ideally, the HLO should demonstrate that things got much faster.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3106,169 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+//   Compute values of slice input operands
+//   Br slice_guard_0
+//
+// slice_guard_0:
```
Sorry for nitpicking, I'm still trying to understand whether we actually need non-structured control flow here.

Is the code snippet below equivalent to

```
if (guarding_cond0) {
  // Write to output of slice0
}
if (guarding_cond1) {
  // Write to output  of slice 1
}
...

?
```

trentlo replied:
Gave it some more thoughts now. Yes, it can be written in a structured control flow. I did not pay too much attention to it as I initially use the concat codegen as my reference point. If this really bothers you, I can try to re-write it into structured control flow. Please let me know what you feel.

cheshire replied:
I do not want to block your work on this, but it seems that structured flow would be easier to read and maintain. Do you want to try it and see whether it is indeed the case? I'm having a hard time following what exactly the emitter is doing due to all the stateful blocks.

trentlo replied:
No problem. Let's me give it a try. I think it will work.

trentlo replied:
Done the rewriting. The code is indeed much cleaner. Thanks for asking the good question.

trentlo commented:
> Thanks, e.g.
> [c6ccf63](https://github.com/tensorflow/tensorflow/commit/c6ccf6316430193e18dde016698d2d8c9989b30a)
> 
> You can get the numbers by running with `XLA_FLAGS=--xla_hlo_profile`
> 
> Ideally, the HLO should demonstrate that things got much faster.

Ah~ I now see where you are from.

Some context: this kind of input-fusible slices is not supported by the current XLA. So, I cannot create a fused computation to do the before- and after-change comparison as you did in the given nice example. This input-fusible slice is needed by the horizontal fusion in this depending [PR](https://github.com/tensorflow/tensorflow/pull/34882). So, you can find more involved tests there (For example RMSPropLike in the [test file](https://github.com/tensorflow/tensorflow/pull/34882/files#diff-b40b77c469fb9627d0c46bf36458ce66)). The slice input fusion is exercised as the horizontal fusion relies on them for fusion. I separated them into two PRs because it will help with the review process. Briefly speaking, the horizontal fusion fuses the outputs of two independent fused computation to reduce the kernel launch overhead.

Please let me know if this fulfills your request about more involved tests. It makes sense to me that the tests that come with this PR are unit-tests while the (more involved) integration tests are with the above-mentioned horizontal fusion PR. Feel free to let me know if this makes sense to you.

In addition, I benchmarked the [tests](https://github.com/tensorflow/tensorflow/pull/34882/files#diff-b40b77c469fb9627d0c46bf36458ce66) with `--xla_hlo_profile`. Logs are here: [log.on.txt](https://github.com/tensorflow/tensorflow/files/3947407/log.on.txt) and [log.off.txt](https://github.com/tensorflow/tensorflow/files/3947408/log.off.txt).

Below is the summary (by greping). I will add them into the commit log of the horizontal fusion. In the GradientDescentOptimizerLike case, the time is reduced from 2.39ms to 311us. Please let me know if this responded to your requests to benchmark numbers.

```
root@686952bd1fe5:/opt/tensorflow/tensorflow-source# grep -r "Execution profile" log.off.txt | grep GradientDescent
2019-12-10 22:05:45.215015: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for GradientDescentOptimizerLike: (2.39 ms @ f_nom)
root@686952bd1fe5:/opt/tensorflow/tensorflow-source# grep -r "Execution profile" log.on.txt | grep GradientDescent
2019-12-10 22:05:03.831600: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for GradientDescentOptimizerLike: (311 us @ f_nom)
root@686952bd1fe5:/opt/tensorflow/tensorflow-source# grep -r "Execution profile" log.off.txt | grep RMSProp
2019-12-10 22:05:48.877372: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for RMSPropLike: (980 us @ f_nom)
root@686952bd1fe5:/opt/tensorflow/tensorflow-source# grep -r "Execution profile" log.on.txt | grep RMSProp
2019-12-10 22:05:13.513901: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for RMSPropLike: (112 us @ f_nom)
```

trentlo committed:
Comment and style polishing.

trentlo reviewed (commented):
Responded.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -301,6 +301,44 @@ llvm::Type* GetIndexTypeForKernel(const HloInstruction* hlo, int64 launch_size,
   return b->getInt32Ty();
 }
 
+// Get the input shape of the ROOT slices, which will be used as the kernel
```
Another nitpick: for function docstrings we use declarative rather than imperative, cf. https://google.github.io/styleguide/cppguide.html#Function_Comments

This is applicable here and below.

trentlo replied:
Got it. Will do.

trentlo replied:
Done.

cheshire reviewed (changes requested):
> So, I cannot create a fused computation to do the before- and after-change comparison as you did in the given nice example

I might be missing something, but your change enables a fusion which was previously impossible, right? So you can profile an HLO which previously did not have this fusion, and the profile should show that it got (at least slightly) faster after this pass, right?

trentlo commented on tensorflow/compiler/xla/service/gpu/ir_emission_utils.cc:
```diff
@@ -234,6 +234,31 @@ bool IsReductionFromOrToContiguousDimensions(const HloInstruction& reduce) {
+    return false;
+  }
+
+  return absl::c_all_of(root->operands(), [&](const HloInstruction* insn) {
```
I use `insn` as the shorthand. Two styles (instr and insn) are both used by people.

trentlo replied:
On a second thought, changed it to `instr` as it is a more widely used style in Tensorflow.

trentlo reviewed (commented):
Responded. Will push a commit soon.

trentlo commented:
> I might be missing something, but your change enables a fusion which was previously impossible, right? So you can profile an HLO which previously did not have this fusion, and the profile should show that it got (at least slightly) faster after this pass, right?

The difficulty I am having is that this commit touches only code generation. However, we cannot have a HLO with input-fusible slices to exercises both the codegen before and after this change, as input-fusible slices are _not_ supported by the existing codegen (without the change.)

Alternatively, we can have an HLO with several kLoop fusions and use [the horizontal fusion pass](https://github.com/tensorflow/tensorflow/pull/34882) to generate input-fusible slices (by fusing the kLoop fusions). We can turn on and off the horizontal fusion pass for comparison; the slice input fusion is generated (after fusing the kLoop fusions) when the horizontal fusion is on, while kLoop fusions remain if the horizontal fusion is off. This demonstrates the effects of the change. The tests are already in the depending [PR](https://github.com/tensorflow/tensorflow/pull/34882) and numbers are reported as in my previous replies.

Please let me know if this addresses your concerns or not.

trentlo reviewed (commented):
Some comments are addressed with the new commit.

trentlo committed:
Polishing comments.

trentlo reviewed (commented):
Have responded all the previous comments. Please help to take a look again when you have a moment. Thanks!

trentlo committed:
Rewrite EmitElementForInputFusibleSlices using structured control flow.

cheshire reviewed (approved):
Thanks, that's much easier to read!

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3106,122 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+  VLOG(10) << "Emitting slice input fusion for " << unnested_hlo->ToString();
+
+  HloInstruction* slice_or_tuple = unnested_hlo->fused_expression_root();
+  auto slice_instructions = [&]() -> absl::Span<HloInstruction* const> {
```
Actually you can just write `[&] -> ...`

trentlo replied:
Ah. Will make the change.

trentlo replied:
Did not make this change as I don't want to drop `-> absl::Span<HloInstruction* const>`

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emission_utils.h:
```diff
@@ -157,6 +157,12 @@ bool ImplementedAsLibraryCall(const HloInstruction& hlo);
 bool IsReductionFromOrToContiguousDimensions(const HloInstruction& reduce);
 
+// Returns whether unnested_hlo is an input fusion whose root is either a slice
+// or a tuple of slices. If verify_no_strides is true, return false unless all
```
"Returns" here as well.

trentlo replied:
Good catch.

trentlo replied:
Done.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3106,122 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+    }
+    llvm::Value* guarding_cond = b_.CreateAnd(index_within_ranges);
+
+    auto emit_slice_elem_func = [&]() -> void {
```
Similarly, brackets after `[&]` can be dropped.

trentlo replied:
Done.

cheshire commented on tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3059,5 +3106,122 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+  UpdateLaunchDimensions(launch_dimensions, kernel_thunk.get(),
+                         ir_emitter_context_->llvm_module());
+
+  auto loop_body_generator =
```
Nitpick: I would just declare the lambda inline inside the call, but up to you.

trentlo replied:
Will move it inside since this is short.

trentlo replied:
Done.

trentlo reviewed (commented):
Thanks for the review! George. I will push a new commit soon.

trentlo reviewed (commented):
Done.

trentlo committed:
Minor coding style polishing.
